### PR TITLE
Feature/grad add data to callback 'on_grad_computed'

### PR DESCRIPTION
**Pull request by spott**
Work towards #295

### CODE REVIEW

**Comment by taketwo** [comment]
Please update documentation [here](https://skorch.readthedocs.io/en/latest/user/callbacks.html#on-grad-computed). Also probably makes sense to update the signature in all classes that have this callback, even though they don't use the new arguments (`NeuralNet`, `GradientNormClipping`, `Callback`).

**Comment by spott** [comment]
This is something that I'm a little bit unclear about:

The `Callback` base class only has `net` and `**kwargs` for all callback functions.  I actually had to dig into the `NeuralNet` class in order to figure out what is passed, and what the names of the parameters are.

I figure this is to allow the callbacks to be flexible when the `NeuralNet` class is subclassed, but maybe it should be documented somewhere other than the source code?

I'm happy to add the calling signature to `Callback`, `NeuralNet` and `GradientNormClipping` if we think that they should be there?

**Comment by BenjaminBossan** [comment]
> I figure this is to allow the callbacks to be flexible when the NeuralNet class is subclassed, but maybe it should be documented somewhere other than the source code?

Yes, this is the reason why we did it this way. It is somewhat documented [here](https://skorch.readthedocs.io/en/latest/user/callbacks.html#callback-base-class) but if you think more documentation is needed, we should add it.

> I'm happy to add the calling signature to Callback, NeuralNet and GradientNormClipping

So far, we haven't added arguments that were not needed in the methods but technically, it would be better if the methods have the same signature, so now might be a good time to start it.

File: docs/user/callbacks.rst [hunk 1]
```diff
@@ -48,44 +48,51 @@ on_train_begin
 ^^^^^^^^^^^^^^
 
 Called once at the start of the training process (e.g. when calling
-fit).
+fit).  Signature: `on_train_begin(net, X, y)`.
```
Comment (taketwo) [inline docs/user/callbacks.rst]: Maybe paste signatures directly into the headers? There are precedents for doing this already: https://skorch.readthedocs.io/en/latest/user/neuralnet.html#fit-x-y

**Comment by benjamin-work** [comment]
@spott give us a ping when this is ready for review

**Comment by spott** [comment]
@benjamin-work. I think this is ready for review.

I've changed the generic callback signatures, but not the signatures for each of the callbacks (for example, `GradientNormClipping`'s `on_grad_computed` callback doesn't accept `Xi` and `yi`, it just accepts `**kwargs`:  I think it is cleaner that way (it is only accepting that which it is using).  If you disagree, I can change it.

`net` and `callbacks/base.py` have been changed to reflect the accepted callback signature.  I have left `**kwargs` in there to allow for flexibility.

**Comment by benjamin-work** [comment]
LGTM, thanks.

**Pull request merged**